src/Auth/Controllers/AccountController.cs

```diff
@@ -0,0 +1,301 @@
+{
+    [SecurityHeaders]
+    [AllowAnonymous]
+    public class AccountController : Controller
+    {
+        private readonly UserManager<ApplicationUser> _userManager;
+        private readonly SignInManager<ApplicationUser> _signInManager;
+        private readonly IIdentityServerInteractionService _interaction;
+        private readonly IClientStore _clientStore;
+        private readonly IAuthenticationSchemeProvider _schemeProvider;
+        private readonly IEventService _events;
+
+        public AccountController(
+            UserManager<ApplicationUser> userManager,
+            SignInManager<ApplicationUser> signInManager,
+            IIdentityServerInteractionService interaction,
+            IClientStore clientStore,
+            IAuthenticationSchemeProvider schemeProvider,
+            IEventService events)
+        {
+            _userManager = userManager;
+            _signInManager = signInManager;
+            _interaction = interaction;
+            _clientStore = clientStore;
+            _schemeProvider = schemeProvider;
+            _events = events;
+        }
+
+        /// <summary>
+        /// Entry point into the login workflow
+        /// </summary>
+        [HttpGet]
+        public async Task<IActionResult> Login(string returnUrl)
+        {
+            // build a model so we know what to show on the login page
+            var vm = await BuildLoginViewModelAsync(returnUrl);
+            return View(vm);
+        }
+
+        /// <summary>
+        /// Handle postback from username/password login
+        /// </summary>
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<IActionResult> Login(LoginInputModel model, string button)
+        {
+            // check if we are in the context of an authorization request
+            var context = await _interaction.GetAuthorizationContextAsync(model.ReturnUrl);
+
+            // the user clicked the "cancel" button
+            if (button != "login")
+            {
+                if (context != null)
+                {
+                    // if the user cancels, send a result back into IdentityServer as if they 
+                    // denied the consent (even if this client does not require consent).
+                    // this will send back an access denied OIDC error response to the client.
+                    await _interaction.GrantConsentAsync(context, ConsentResponse.Denied);
+
+                    // we can trust model.ReturnUrl since GetAuthorizationContextAsync returned non-null
+                    if (await _clientStore.IsPkceClientAsync(context.ClientId))
+                    {
+                        // if the client is PKCE then we assume it's native, so this change in how to
+                        // return the response is for better UX for the end user.
+                        return View("Redirect", new RedirectViewModel { RedirectUrl = model.ReturnUrl });
+                    }
```

Comment:
You are suggesting I remove the whole if because a line letter it says `Redirect(model.ReturnUrl)`.
This JavaScript is better in case the one authenticating is a native mobile application its not exactly as simple as redirect:
`<meta http-equiv="refresh" content="0;url=@Model.RedirectUrl" data-url="@Model.RedirectUrl">
`
I have to be honest, I am not very sure how this is different and how it works for native mobile apps but I want to trust IdentityServer that they knew what they were doing

Reply:
Personally, the code shouldn't exist if we can't explain why it's needed. I would suggest removing it and if it breaks later we can re-add the code with a comment explaining why it's needed.

Reply:
Well, I went and checked it more.
As I understood it this moving from page to page is part of https://tools.ietf.org/html/rfc7636
The 302 that Redirect does is incorrect and is unsafe for injections.
Also extra reading: https://christianlydemann.com/implicit-flow-vs-code-flow-with-pkce/

Reply:
> Since the access token is sent as the fragment (hash) part of the redirect URL (also called the front channel), it will be exposed to an attacker if the redirect is intercepted. Now, of course, it is recommended to always use HTTPS, but even HTTPS can be compromised by man-in-the-middle attacks with methods as corporate SSL certificates to inspect traffic.
  >
  > This threat is mitigated in code flow by not getting the token response from a redirect but instead a direct request-response HTTP request.

I find this very questionable, because if a redirect was compromised, why not the normal request too? I wasn't able to find anything else that supports that.

However, most of these stories have some sort of truth to them. Personally, the best action would be to keep the code and add a comment to that article.

Reply:
I think I found the problem this is based on:

  > ## An important note about referers on older browsers
  > 
  > Even though HTTPS encrypts the path you are accessing, if you click a
  > hyperlink within that site which goes to an unencrypted page, the full path
  > may be leaked in the `referer` header. This is [not the case anymore][1] for
  > many newer browsers, but older or non-compliant browsers may still have this
  > behavior, as will websites which set the HTML5 referer meta tag to always
  > send the information. An example sent *unencrypted* by a client to jump from
  > `https://example.com/private/details.html` to
  > `http://example.org/public/page.html` in such a case would be:
  > 
  > ~~~none
  > GET /public/page.html
  > Referer: https://example.com/private/details.html
  > User-Agent: Wget/1.19.1 (linux-gnu)
  > Accept: */*
  > Accept-Encoding: identity
  > Host: example.org
  > ~~~ 
  > 
  > As such, navigating from an HTTPS page to an HTTP page may leak the full URL
  > (excluding the fragment) of the previous page, so keep that in mind.
  > <sub>[(source)][2]</sub>

However, even this source says that 

  > As such, navigating from an HTTPS page to an HTTP page may leak the full URL
  > **(excluding the fragment)** of the previous page, so keep that in mind.
  > <sub>[(source)][2]</sub>

which contradicts

  > Since **the access token is sent as the fragment** (hash) part of the redirect URL (also called
  > the front channel), it will be exposed to an attacker if the redirect is intercepted.
  > <sub>[(source)][3]</sub>

  [1]: http://smerity.com/articles/2013/where_did_all_the_http_referrers_go.html
  [2]: https://security.stackexchange.com/a/176169/223751
  [3]: https://christianlydemann.com/implicit-flow-vs-code-flow-with-pkce/

Reply:
I figured it out, [it's for user experience][1].

  [1]: https://github.com/IdentityServer/IdentityServer4.Demo/blob/1f98c8012d7ff08c9a198660f3fe6ad8dafc6002/src/IdentityServer4Demo/Quickstart/Account/AccountController.cs#L95-L96

Reply:
Alright, I suppose. We could technically remove it but I guess better UX is ok